### PR TITLE
patch S3 storage basic

### PR DESCRIPTION
**Pull request by white-cc**
### Changelog category :
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

patch S3 storage for ByConity, it will be a preview feature in GA release.
- [x] read
- [x] write
- [x] merge

Attach and others will be in another request.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by hustnn** [comment]
Thanks @white-cc . 

@Clark0 @linkwk7 Pls help to review

File: src/Disks/IDisk.cpp [hunk 1]
```diff
@@ -15,6 +15,8 @@ namespace ErrorCodes
     extern const int NOT_IMPLEMENTED;
 }
 
+std::atomic<UInt64> next_disk_id = 0;
```
Comment (linkwk7) [inline src/Disks/IDisk.cpp:18]: seem not used in this path?

File: src/Storages/S3/RAReadBufferFromS3.cpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    if (outcome.IsSuccess())
+    {
+        *reader = outcome.GetResultWithOwnership();
+        reader_end_offset += reader->GetContentLength();
+
+        return true;
+    }
+    else
+    {
+        if (outcome.GetError().GetExceptionName() == "InvalidRange")
```
Comment (linkwk7) [inline src/Storages/S3/RAReadBufferFromS3.cpp:104]: can be replaced by outcome.GetError().GetResponseCode() == Aws::Http::HttpResponseCode::REQUESTED_RANGE_NOT_SATISFIABLE

File: src/Disks/IDisk.cpp (hunk 1 above)
Comment (white-cc) [reply src/Disks/IDisk.cpp:18]: For DiskByteS3 Obtain disk_id.

File: src/Disks/DiskByteS3.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
+#include <filesystem>
```
Comment (WillemJiang) [inline src/Disks/DiskByteS3.cpp:1]: Please add the license header just like this 
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/Disks/DiskByteS3.h [hunk 4]
```diff
@@ -0,0 +1,125 @@
+#pragma once
```
Comment (WillemJiang) [inline src/Disks/DiskByteS3.h:1]: Please add the license header just like this
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/Disks/DiskType.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#include <Disks/DiskType.h>
```
Comment (WillemJiang) [inline src/Disks/DiskType.cpp:1]: Please add the license header just like this
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/IO/LimitSeekableReadBuffer.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
+#include <IO/LimitSeekableReadBuffer.h>
```
Comment (WillemJiang) [inline src/IO/LimitSeekableReadBuffer.cpp:1]: Please add the license header just like this
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/IO/LimitSeekableReadBuffer.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
+#pragma once
```
Comment (WillemJiang) [inline src/IO/LimitSeekableReadBuffer.h:1]: Please add the license header just like this
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/Storages/MergeTree/S3ObjectMetadata.cpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
+#include <Storages/MergeTree/S3ObjectMetadata.h>
```
Comment (WillemJiang) [inline src/Storages/MergeTree/S3ObjectMetadata.cpp:1]: Please add the license header just like this
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/Storages/S3/RAReadBufferFromS3.cpp [hunk 9]
```diff
@@ -0,0 +1,249 @@
+#include <Storages/S3/RAReadBufferFromS3.h>
```
Comment (WillemJiang) [inline src/Storages/S3/RAReadBufferFromS3.cpp:1]: Please add the license header just like this
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/Storages/S3/RAReadBufferFromS3.h [hunk 10]
```diff
@@ -0,0 +1,91 @@
+#pragma once
```
Comment (WillemJiang) [inline src/Storages/S3/RAReadBufferFromS3.h:1]: Please add the license header just like this
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/Storages/S3/WriteBufferFromByteS3.cpp [hunk 11]
```diff
@@ -0,0 +1,317 @@
+#include <Storages/S3/WriteBufferFromByteS3.h>
```
Comment (WillemJiang) [inline src/Storages/S3/WriteBufferFromByteS3.cpp:1]: Please add the license header just like this
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/Storages/S3/WriteBufferFromByteS3.h [hunk 12]
```diff
@@ -0,0 +1,99 @@
+#pragma once
```
Comment (WillemJiang) [inline src/Storages/S3/WriteBufferFromByteS3.h:1]: Please add the license header just like this
https://bytedance.feishu.cn/docx/MtIsdOS7coFyjJxVAlncHNg8n9f
if the code is wroten from scratch.

File: src/Disks/DiskType.cpp [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+int DiskType::toInt(Type disk_type)
+{
+    return static_cast<int>(disk_type);
+}
+
+DiskType::Type DiskType::toType(int disk_type_id)
+{   
+    if (disk_type_id > 5)
```
Comment (Clark0) [inline src/Disks/DiskType.cpp:19]: The magic number is error-pone, better to add a enum SIZE in Type

File: src/Disks/IDisk.cpp (hunk 1 above)
Comment (Clark0) [reply src/Disks/IDisk.cpp:18]: should it be a static variable?

File: src/Storages/S3/ReadBufferFromByteS3.h [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+{
+class S3Client;
+}
+
+namespace DB
+{
+/**
+ * Perform S3 HTTP GET request and provide response to read.
+ */
+class ReadBufferFromByteS3 : public ReadBufferFromFileBase
```
Comment (hustnn) [inline src/Storages/S3/ReadBufferFromByteS3.h:21]: Is it being used?

File: src/Disks/IDisk.cpp (hunk 1 above)
Comment (white-cc) [reply src/Disks/IDisk.cpp:18]: static declaration of follows non-static declaration. 
I use ‘extern’ to solve use in other file.

**Review by hustnn — APPROVED** [review]

**Pull request merged**